content/drafts/sphinx.md

```diff
@@ -0,0 +1,447 @@
+then you are once again left with the clear text. The sphinx packet 
+construction makes heavy use of XOR for encrypting a clear text packet using a 
+pseudo random byte stream.
+
+# Sender preparation
+
+The example used in this post continues where the last post ended off. 
+Basically we have the following route where Alice is the sender and Dave is the 
+recipient. Alice essentially just needs to deliver a payload to each hop 
+without leaking too much data about the route.
+
+![](/onion/1-path.png#center)
+
+Alice has a set of payloads that she wants to communicate to each hop. When the 
+payloads are put into the onion packet, they are prefixed with a length and 
+postfixed with a 32 HMAC. 
+
+![](/onion/5-payloads.png#center)
+
+Alice now goes ahead and generates a `session_key` and uses this to derive the 
+chain of [ephemeral keys](#ephemeral-keys). 
+
+![](/onion/3-ephemeral_keys.png#center)
+
+I mentioned before that we don’t want hops on the route to have an idea of where 
+on the route they are. To achieve this, the onion packet given to each hop will 
+be the exact same size (1300 bytes) and each hop will only be able to read the 
+payload meant for it. The rest of the onion will look like a random stream of 
+bytes to that hop.
+
+# First attempt at wrapping the onion. 
+
+We are going to run through the onion wrapping process twice. During the first 
+pass we will get an initial feel for it and we will see why certain complexities 
+need to be added on the second pass.
+
+Alice needs to wrap the onion back to front: we first add the payload for Dave, 
+encrypt that and then add the payload for Charlie and so on. The packet we end 
+up giving to Bob will be a fully wrapped onion. Each hop will peel one layer of 
+the onion.
+
+First, Alice generates a 1300 byte pseudo random stream using the session key. 
+This is knows as the "padding". 
+
+![](/onion/6-padding.png#center)
+
+### Wrapping for Dave
+
+She then slides over the padding to make room for Dave’s payload. The HMAC that 
+is appended to Dave’s payload is not actually a real HMAC. Since Dave is the 
+last hop, this onion doesn’t need to be passed onto any other hops and so this 
+HMAC is made to be an empty set of zero bytes which serves as a signal to Dave 
+that he is the final hop on the route.
+
+The packet needs to remain 1300 bytes long and so that trailing section of the 
+padding is chopped off.
+
+![](/onion/7-wrap_dave_1.png#center)
+
+Alice then uses her shared key with Dave, $ss_{AB}$ along with the `rho` 
+constant to derive a pseudo random stream of 1300 bytes. This stream is XOR'd 
+with the onion packet which produces an encrypted packet that only Dave would 
+be able to decrypt. Alice uses the $e_{AD}$ ephemeral key along with a hash of
+the onion at this point to calculate an HMAC for this packet. Dave will later 
+verify that his HMAC is valid against the hash of the packet he receives along 
+with the ephemeral public key $E_{AD}$. 
```

Comment:
See my comment above about HMACs and private keys... So I don't think this is correct? Can learn Dave $e_{AD}$ somehow during the decryption to verify the HMAC here?

Reply:
Or is $E_{AD}$ used for the HMAC? And only the person knowing the actual ephemeral public key can create the HMAC, proving it was encrypted for the intended recipient?

Reply:
updated all the HMAC text!